docs/roadmap/scaling-2019.md

```diff
@@ -0,0 +1,115 @@
+
+1. identify and programatically capture sources of cold-start latency at all levels of the stack ([#2495](https://github.com/knative/serving/issues/2495))
+2. chase down the low hanging fruit (e.g. [#2659](https://github.com/knative/serving/issues/2659))
+3. architect solutions to larger chunks of cold-start latency
+
+**Our goal is to achieve sub-second average cold-starts by the end of the year.**
+
+* POC: Greg Haynes (IBM)
+* Github: [Project 8](https://github.com/knative/serving/projects/8)
+
+### Overload Handling
+
+Knative Serving provides concurrency controls to limit the number of requests a container must handle simultaneously. Additionally, each pod has a queue for holding requests when the container concurrency limit has been reached. When the pod-level queue overflows, subsequent request are rejected with 503 "overload".
+
+This is desirable to protect the Pod from being overloaded. But in the aggregate the behavior is not ideal for situations when autoscaling needs some time to react to sudden increases in request load.  This could be when the revision is scaled to zero.  Or when the revision is already running some pods, but not nearly enough.
+
+The goal of Overload Handling is to enqueue requests at a revision-level. Scale-from-zero should not overload if autoscaling can react in a reasonable amount of time to provide additional pods. When new pods come online, they should be able to take load for the existing pods. Even when scaled above zero, brief spikes of overload should be handled by enqueuing requests at a revision-level. The depth of the revision-level queue should also be configurable because even the Revision as a whole needs to guard against overload.
+
+The overall problem touches on both Networking and Autoscaling, two different working groups. Much of the overload handling will be implemented in the Activator, which is a part of ingress. So this project is shared jointly between the two working groups.
+
+**Our goal is for a Revision with 1 Pod and a container concurrency of 1 to handle 1000 requests arriving simultaneously within 30 seconds with no errors, where each request sleeps for 100ms.**
+
+E.g.
+```yaml
+apiVersion: serving.knative.dev/v1alpha1
+kind: Service
+metadata:
+  name: overload-test
+spec:
+  runLatest:
+    configuration:
+      revisionTemplate:
+        metadata:
+          annotations:
+            autoscaling.knative.dev/minScale: "1"
+            autoscaling.knative.dev/maxScale: "10"
+        spec:
+          containerConcurrency: 1
+          container:
+            image: gcr.io/joe-does-knative/sleep:latest
+            env:
+            - name: SLEEP_TIME
+              value: "100ms"
+```
+```bash
+for i in `seq 1 1000`; do curl http://$MY_IP & done
+```
+
+This verifies that 1) we can handle all requests in an overload without error and 2) all the requests don't land on a single pod, which would take 100 sec.
```

Comment:
There's an assumption that `containerConcurrency` works correctly & doesn't regress. 

Should we validate the # of requests a pod receives via a count that we query after the fact?

Reply:
We should have an E2E test for `containerConcurrency` as well.